File: LeetCode-Java/src/Tree/LeetCode/LeetCode105.java
package Tree.LeetCode;

import Basic.string.StringAlgo;
import jdk.nashorn.api.tree.Tree;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * 根据一棵树的前序遍历与中序遍历构造二叉树。
 * <p>
 * 注意:
 * 你可以假设树中没有重复的元素。
 */
public class LeetCode105 {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return _buildTree(preorder, 0, inorder);
    }

    private TreeNode _buildTree(int[] preorder, int preorderRootIndex, int[] inorder) {
        if (inorder.length == 0) {
            return null;
        }
        if (inorder.length == 1) {
            return new TreeNode(inorder[0]);
        }
        final TreeNode rootNode = new TreeNode(preorder[preorderRootIndex]);

        /// 从中序遍历中到当前根节点，当前根节点左边的元素是左子树，右边的元素是右子树
        int rootIndex = 0;
        for (int i = 0; i < inorder.length; i++) {
            /// 找到根节点
            if (inorder[i] == preorder[preorderRootIndex]) {
                rootIndex = i;
                break;
            }
        }
        int[] leftNodes = Arrays.copyOfRange(inorder, 0, rootIndex);
        int[] rightNodes = Arrays.copyOfRange(inorder, rootIndex + 1, inorder.length);
        rootNode.left = _buildTree(preorder, preorderRootIndex + 1, leftNodes);
        rootNode.right = _buildTree(preorder, preorderRootIndex + leftNodes.length + 1, rightNodes);
        return rootNode;
    }

    public static void main(String[] args) {
        LeetCode105 leetCode105 = new LeetCode105();
        TreeNode node = leetCode105.buildTree(new int[]{3,9,8}, new int[]{8,9,3});
        System.out.println(node);
    }
}

File: LeetCodeSwift/LeetCodeSwift/Tree/LeetCode105.swift
//
//  LeetCode105.swift
//  LeetCodeSwift
//
//  Created by Liang on 2020/5/22.
//  Copyright © 2020 imieutan. All rights reserved.
//

import Foundation
/**
 105. 从前序与中序遍历序列构造二叉树
 根据一棵树的前序遍历与中序遍历构造二叉树。

 注意:
 你可以假设树中没有重复的元素。

 例如，给出

 前序遍历 preorder = [3,9,20,15,7]
 中序遍历 inorder = [9,3,15,20,7]
 返回如下的二叉树：

     3
    / \
   9  20
     /  \
    15   7
 */
class LeetCode105 {
    func buildTree(_ preorder: [Int], _ inorder: [Int]) -> TreeNode? {
        if preorder.count == 0 {
            return nil
        }
        return buildTree(preorder, currentPreorderIndex: 0, inorder: inorder)
    }
    
    private func buildTree(_ preorder: [Int], currentPreorderIndex: Int, inorder: [Int]) -> TreeNode? {
        if inorder.count == 0 || currentPreorderIndex >= preorder.count {
            return nil
        }
        let rootVal = preorder[currentPreorderIndex];
        let rootNode = TreeNode(rootVal)
        let rootIndex = findRootValue(rootVal, inorder: inorder)
        
        let leftNode = buildTree(preorder, currentPreorderIndex: currentPreorderIndex + 1, inorder: Array(inorder[0..<rootIndex]))
        /// 前序遍历中右子树的起点是 rootIndex + left Tree node counts
        let rightNode = buildTree(preorder, currentPreorderIndex: currentPreorderIndex + rootIndex + 1, inorder: Array(inorder[rootIndex + 1..<inorder.count]))
        rootNode.left = leftNode
        rootNode.right = rightNode
        return rootNode
    }
    
    private func findRootValue(_ rootValue: Int, inorder: [Int]) -> Int {
        for i in 0..<inorder.count {
            if rootValue == inorder[i] {
                return i
            }
        }
        return -1
    }
    
}
